stm32: clock_control: Enforce HCLK prescaler value

erwango opened the pull request:
STM32 clock control subsystem allows to configure a different
frequency value for core clock (SYSCLK) and AHB clock (HCLK).
Though, it is HCLK which is used to feed Cortex Systick timer
which  is used in zephyr as reference system clock.
If HCLK frequency is configured to a different value from SYSCLK
frequency, whole system is exposed to desynchro between zephyr clock
subsytem and STM32 HW configuration.
To prevent this, and until zephyr clock subsystem is changed to be
aware of this potential configuration, enforce AHB prescaler value
to 1 (which is current default value in use for all STM32 based
boards).

On STM32H7, enforce D1CPRE which fills the same role as ABH precaler.

On STM32MP1, the equivalent setting is done on A7 core, so it is
not exposed to the same issue as long as SYS_CLOCK_HW_CYCLES_PER_SEC
is set with the 'mlhclk_ck' clock frequency value. Update
matching boards documentation.

Fixes #17188

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Adding the backport tag since this is valid on 1.14 branch as well.
Though, commit will have to be reworked since STM32H7 is not supported on 1.14

dbkinder commented on boards/arm/96b_avenger96/doc/index.rst:
```diff
@@ -185,7 +185,8 @@ Default Zephyr Peripheral Mapping:
 
-The Cortex®-M4 Core is configured to run at a 209 MHz clock speed.
+The Cortex®-M4 Core is configured to run at a 209 MHz clock speed. This value
+should match the configured mlhclk_ck frequency.
```
should this be ``must match`` (is it OK if it doesn't?)

erwango replied:
it must indeed, txs

dbkinder commented on boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst:
```diff
@@ -196,7 +196,8 @@ Default Zephyr Peripheral Mapping:
 
-The Cortex®-M4 Core is configured to run at a 209 MHz clock speed.
+The Cortex®-M4 Core is configured to run at a 209 MHz clock speed. This value
+should match the configured mlhclk_ck frequency.
```
Same here, should this be ``must match``?

erwango committed:
stm32: clock_control: Enforce HCLK prescaler value

STM32 clock control subsystem allows to configure a different
frequency value for core clock (SYSCLK) and AHB clock (HCLK).
Though, it is HCLK which is used to feed Cortex Systick timer
which  is used in zephyr as reference system clock.
If HCLK frequency is configured to a different value from SYSCLK
frequency, whole system is exposed to desynchro between zephyr clock
subsytem and STM32 HW configuration.
To prevent this, and until zephyr clock subsystem is changed to be
aware of this potential configuration, enforce AHB prescaler value
to 1 (which is current default value in use for all STM32 based
boards).

On STM32H7, enforce D1CPRE which fills the same role as ABH precaler.

On STM32MP1, the equivalent setting is done on A7 core, so it is
not exposed to the same issue as long as SYS_CLOCK_HW_CYCLES_PER_SEC
is set with the 'mlhclk_ck' clock frequency value. Update
matching boards documentation.

Fixes #17188

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@dbkinder, doc updated, txs for the review

dbkinder reviewed (approved):
+1 for doc changes, thanks.
